### DISCUSSION TITLE
Export empty namespace

### DISCUSSION THREAD

**Discussion by roin93** [FreeOpcUa/opcua-asyncio#667] [Q&A] [answered]
Hey guys,

I create a server like the following:

```
server = Server()
await server.init()
server.set_endpoint("opc.tcp://%s:%s/opcua/" % (ip_addr, port))
server.set_server_name("MyServer")

mynamespace1 = await server.register_namespace("mynamespace1") 
mynamespace2 = await server.register_namespace("mynamespace2") 

someObj = await server.nodes.objects.add_object(mynamespace2, "TestObj")  # Attention, namespace1 is missing
```

After creation I export it via `server.export_xml_by_ns("nodes.xml", ["mynamespace1", "mynamespace2"])`

Problem: The xml file does not contain the (currently) empty namespace "mynamespace1". When I import the xml via `server.import_xml("nodes.xml")` and also import some reference files, which reference to the node id (e.g. 'ns=2;i=100'), my whole project does not work because the nodeids does not match anymore.

Is there any way to export empty namespaces?

**Comment by xydan83** [comment]
Based on the code, if I'm not mistaken, at the moment the namespaces are formed in the "**build_etree**" method and added from the list of nodes, in the "**_add_namespaces**" method, therefore, if there is no node, there is no namespace.
![image](https://user-images.githubusercontent.com/86729310/135426643-7c8ef925-af7b-44c2-8ac0-0e960b01f94a.png)

You can try adding one fictitious node as an object, maybe this option will suit you?

**Comment by oroulet** [comment]
yes yo need to change export code. But does it really make sense to export namespaces without nodes?

**Reply by roin93** [reply]
As I mentioned I try to mirror a server. Starting the mirroring script takes somes time so I want to speed things up by saving the current state of my server (via the export function).
If there are _currently_ empty namespaces, I need to keep them, especially for the correct references. A missing namespace would cause my references to get broken as the identifiers could be changed or cannot be created the same way as i exported them.

**Comment by xydan83** [comment]
Now I also ran into this problem. It looks like we need to slightly change the method for adding namespase, since if you make a mirror from servers, then the parsers change the identifiers in the nodes :(

**Comment by xydan83** [comment, marked as answer]
Alternatively, you can override the _add_namespaces method, remove the filter checking for nodes from it.

```python
    async def _add_namespaces(self, nodes):
        ns_to_export: list = (await self.server.get_namespace_array())[1:]  # ignore OPC Foundation namespace

        # write namespaces to xml
        self._add_namespace_uri_els(ns_to_export)
```

**Reply by roin93** [reply]
Thats not a bad idea. 
Is there a way to add an optional parameter to the export functions to include_empty_namespaces?

**Reply by xydan83** [reply]
I see two ways:
1) Make a PR and offer to inject such a parameter into the project;
2) You can inherit from the XmlExporter class and override the _add_namespaces method. And then work with the export through your class. There will be a workaround.
I went with the second scenario and it works)
I'm not sure that the developers of the library would consider it correct to skip empty namespaces without nodes in the exporter, but I did not see anything prohibiting these actions in the specification for the OPC UA XML.
Maybe I'm missing something.

https://reference.opcfoundation.org/v104/Core/docs/Part6/F.15/
https://reference.opcfoundation.org/Core/docs/Part6/#5.3
